src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -192,41 +58,181 @@
-   - create the list of pivot row value combinations. We don't need the nils group this time; it would result in
-     a doubling of the Grand totals row
-     We DO want to generate tuples that can have 'nil' (eg. ['CA' nil]) so that Row totals rows will be grabbed from the map too.
-   - construct each row by concatenating the row values, since they're the labels for each pivot-row. Handily, the `ks` are exactly
-     these values, so we can just concat each k to its `row-groups` entry
-   - Run the `totals-row-fn` to add the Row totals and Grand totals labels in the right spots."
-  [rows {:keys [pivot-rows] :as pivot-spec} :- ::pivot-spec]
-  (let [row-groups (row-grouper rows pivot-spec)
-        ks         (if (seq pivot-rows)
-                     (mapv vec (concat
-                                (apply math.combo/cartesian-product (map #(all-values-for rows % true) pivot-rows))))
-                     [[nil]])]
-    (->> (map (fn [k] (vec (concat k (get row-groups k)))) ks)
-         (filter #(< (count pivot-rows) (count %)))
-         (map #(totals-row-fn % pivot-spec)))))
-
-(defn- clean-row
-  [row]
-  (mapv #(if (= "" %)
-           nil
-           %)
-        row))
-
-(mu/defn pivot-builder
-  "Create the rows for a pivot export from raw pivot rows `rows`."
-  [rows pivot-spec :- ::pivot-spec]
-  (let [rows       (mapv clean-row rows)
-        pivot-spec (add-pivot-measures pivot-spec)]
-    (vec (concat
-          (header-builder rows pivot-spec)
-          (row-builder rows pivot-spec)))))
+  (-> pivot-spec
+      (assoc :pivot-measures (pivot-measures pivot-spec))
+      (assoc :pivot-grouping (pivot-grouping-key (:column-titles pivot-spec)))))
+
+(mu/defn add-totals-settings :- ::pivot-spec
+  "Given a pivot-spec map without the `:pivot-measures` key, determine what key(s) the measures will be and assoc that value into `:pivot-measures`."
+  [pivot-spec :- ::pivot-spec viz-settings]
+  (let [row-totals (if (contains? viz-settings :pivot.show_row_totals)
+                     (:pivot.show_row_totals viz-settings)
+                     true)
+        col-totals (if (contains? viz-settings :pivot.show_column_totals)
+                     (:pivot.show_column_totals viz-settings)
+                     true)]
+    (-> pivot-spec
+        (assoc :row-totals? row-totals)
+        (assoc :col-totals? col-totals))))
+
+(mu/defn init-pivot
+  "Initiate the pivot data structure."
+  [pivot-spec :- ::pivot-spec]
+  (let [{:keys [pivot-rows pivot-cols pivot-measures]} pivot-spec]
+    {:config         pivot-spec
+     :data           {}
+     :row-values     (zipmap pivot-rows (repeat (sorted-set)))
+     :column-values  (zipmap pivot-cols (repeat (sorted-set)))
+     :measure-values (zipmap pivot-measures (repeat (sorted-set)))}))
+
+(defn- update-set [m k v]
+  (update m k conj v))
+
+(defn- update-aggregate [measure-aggregations new-values agg-fns]
+  (into {}
```

Comment:
also add a docstring explaining this, it's a little confusing

Reply:
I actually am wondering now as I do this if there may be a missed case or two with aggregation; supposing someone sets up an average, the aggregation logic works fine for each 'cell' of the table (since the pivot query processor already does the aggregating), but might not apply when doing totals/subtotals, since the data structure used for the pivot post processing does indeed take care of totals aggregating. This means it might mis-calculate the totals for average. I will check this